File: solutions/javascript/roman-numerals/roman-numerals.js
//
// Given a positive number, convert it into roman numeral string representation.
//
export const toRoman = (number) => {
  if (number < 1 || number > 3999) {
    throw new Error("Number has to be positive integer in range of 1-3999.");
  }

  const ones = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"];
  const tens = ["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"];
  const hund = ["", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"];
  const thou = ["", "M", "MM", "MMM"];

  return (
    thou[Math.floor(number / 1000)] +
    hund[Math.floor((number % 1000) / 100)] +
    tens[Math.floor((number % 100) / 10)] +
    ones[number % 10]
  );
};

File: solutions/cpp/roman-numerals/roman_numerals.cpp
#include "roman_numerals.h"

namespace roman_numerals
{
    std::string thou[10] = {"", "M", "MM", "MMM"};
    std::string hund[10] = {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"};
    std::string tens[10] = {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "IC"};
    std::string ones[10] = {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"};

    std::string convert(int num)
    {
        return thou[num / 1000] +
               hund[num % 1000 / 100] +
               hund[num % 100 / 10] +
               ones[num % 10];
    }
} // namespace roman_numerals
